stackslib/src/chainstate/coordinator/mod.rs

```diff
@@ -2251,40 +2260,54 @@ impl<
         if target_epoch.epoch_id < StacksEpochId::Epoch30 {
             // burnchain has not yet advanced to epoch 3.0
-            self.handle_new_epoch2_burnchain_block(&mut HashSet::new())
+            return self.handle_new_epoch2_burnchain_block(&mut HashSet::new());
         }
-        else {
-            // burnchain has advanced to epoch 3.0, but has our sortition DB?
-            let canonical_snapshot = match self.canonical_sortition_tip.as_ref() {
-                Some(sn_tip) => SortitionDB::get_block_snapshot(self.sortition_db.conn(), sn_tip)?
-                    .expect(&format!(
-                        "FATAL: do not have previously-calculated highest valid sortition tip {}",
-                        sn_tip
-                    )),
-                None => SortitionDB::get_canonical_burn_chain_tip(&self.sortition_db.conn())?,
-            };
-            let target_epoch_index = StacksEpoch::find_epoch(&epochs, canonical_snapshot.block_height).expect("FATAL: epoch not defined for BlockSnapshot height");
-            let target_epoch = epochs.get(target_epoch_index).expect("FATAL: StacksEpoch::find_epoch() returned an invalid index");
 
-            if target_epoch.epoch_id < StacksEpochId::Epoch30 {
-                // need to catch the sortition DB up
-                self.handle_new_epoch2_burnchain_block(&mut HashSet::new())?;
-            }
+        // burnchain has advanced to epoch 3.0, but has our sortition DB?
+        let canonical_snapshot = match self.canonical_sortition_tip.as_ref() {
+            Some(sn_tip) => SortitionDB::get_block_snapshot(self.sortition_db.conn(), sn_tip)?
+                .expect(&format!(
+                    "FATAL: do not have previously-calculated highest valid sortition tip {}",
+                    sn_tip
+                )),
+            None => SortitionDB::get_canonical_burn_chain_tip(&self.sortition_db.conn())?,
+        };
+        let target_epoch_index = StacksEpoch::find_epoch(&epochs, canonical_snapshot.block_height)
+            .expect("FATAL: epoch not defined for BlockSnapshot height");
+        let target_epoch = epochs
+            .get(target_epoch_index)
+            .expect("FATAL: StacksEpoch::find_epoch() returned an invalid index");
 
-            // proceed to process sortitions in epoch 3.0
-            self.handle_new_nakamoto_burnchain_block()
+        if target_epoch.epoch_id < StacksEpochId::Epoch30 {
+            // need to catch the sortition DB up
+            self.handle_new_epoch2_burnchain_block(&mut HashSet::new())?;
         }
+
+        // proceed to process sortitions in epoch 3.0
+        self.handle_new_nakamoto_burnchain_block()
+            .map(|can_proceed| {
+                if can_proceed {
+                    None
+                } else {
+                    // missing PoX anchor block, but unlike in 2.x, we don't know what it is!
+                    Some(BlockHeaderHash([0x00; 32]))
```

Comment:
I think it'd be better to just expand the return type of this function:

```rust
pub enum NewBurnchainBlock {
    ReadyForStacksBlocks,
    WaitForPox2x(BlockHeaderHash),
    WaitForPoxNakamoto,
}

 pub fn handle_new_burnchain_block(&mut self) -> Result<NewBurnchainBlock, Error>
```

That way there's no possibility of confusing the return behavior difference between 2.x and 3.x.

Reply:
Sure